### ISSUE TITLE
Linking in VxWorks failed

### ISSUE THREAD

**Issue by cochicde** [eclipse-paho/paho.mqtt.c#318]
@rpoisel did a great job porting to VxWorks. I tried to compiled it, but I got the error:

ld.exe: -r and -shared may not be used together

This occurs when linking libpaho-mqtt3a.so. I actually just want the static library, but I cannot disable the shared unless touching the CMakeLists.txt

So, my question to @rpoisel is: Did you change something in the CMakeLists.txt file in order to compile it?
P.S.: I added the following flags in CMake:
-D_WRS_KERNEL -DCPU=_VX_ATOM -D_VX_CPU=_VX_ATOM -DTOOL_FAMILY=gnu -DTOOL=gnu -DIP_PORT_VXWORKS=69 -D_VSB_CONFIG_FILE=\"SOME_FOLDER/vsbConfig.h\" -isystem SOME_FOLDER -isystem SOME_FOLDER

**Comment by rpoisel** [comment]
Hello,

we are working with CMake toolchain files. This is what we use to compile for VxWorks 6.9 (from Windows):

```CMake
set(WIN32 0)
set(CMAKE_SYSTEM_NAME VxWorks)
set(CMAKE_SYSTEM_VERSION 1)

list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/../../Modules")

include(${CMAKE_CURRENT_LIST_DIR}/../common/toolchain.env.cmake)

# Name of the target platform
set(CMAKE_SYSTEM_NAME VxWorks)

# Version of the system
set(CMAKE_SYSTEM_VERSION 1)

# specify the cross compiler
if(CMAKE_HOST_SYSTEM_NAME MATCHES "Windows")
    set(CMAKE_C_COMPILER "${COMPILER_HOME}/Wrswrb33_694/gnu/4.3.3-vxworks-6.9/x86-win32/bin/ccpentium.bat")
    set(CMAKE_CXX_COMPILER "${COMPILER_HOME}/Wrswrb33_694/gnu/4.3.3-vxworks-6.9/x86-win32/bin/c++pentium.bat")

    set(TARGET_H_DIR "${COMPILER_HOME}/Wrswrb33_694/vxworks-6.9/target/h")
else()
    message(FATAL_ERROR Unsupported build platform.)
endif()

# compiler/linker flags
set(CMAKE_C_FLAGS
    "--std=gnu99 -Werror=implicit-function-declaration -fno-feature-proxy -O2 -fno-defer-pop -nostdlib -fno-builtin -mno-sse -mno-sse2 -mtune=pentium -march=pentium -ansi -Wall -MD -MP \
    -Wno-strict-aliasing \
    -DCPU=PENTIUM4 -DTOOL_FAMILY=gnu -DTOOL=gnu -D_WRS_KERNEL \
    -I${TARGET_H_DIR} -I${TARGET_H_DIR}/wrn/coreip -I${TARGET_H_DIR}/wrn/coreip/sys"
    CACHE STRING "The compiler flags for compiling C sources" FORCE
)
set(CMAKE_SHARED_LINKER_FLAGS
    "-nostdlib -r -pipe -Wl,-X -T ${TARGET_H_DIR}/tool/gnu/ldscripts/link.OUT"
    CACHE STRING "Additional compiler flags for building shared libraries" FORCE
)
set(CMAKE_EXE_LINKER_FLAGS
    "-fno-feature-proxy -r -pipe -Wl,-X -T ${TARGET_H_DIR}/tool/gnu/ldscripts/link.OUT"
    CACHE STRING "Linker flags to be used to create executables" FORCE
)

set(CMAKE_C_COMPILE_OBJECT "<CMAKE_C_COMPILER> <DEFINES> <INCLUDES> ${CMAKE_C_FLAGS} -o <OBJECT> -c <SOURCE>")
SET(CMAKE_C_CREATE_SHARED_LIBRARY "<CMAKE_C_COMPILER> <LINK_FLAGS> -o <TARGET> <OBJECTS>")
SET(CMAKE_C_LINK_EXECUTABLE "<CMAKE_C_COMPILER> <LINK_FLAGS> <OBJECTS> -o <TARGET>")
```
This is our `toolchain.env.cmake`:
```CMake
if(CMAKE_HOST_SYSTEM_NAME MATCHES "Windows")
#hide sh.exe program
    set(CMAKE_SH CMAKE_SH-NOTFOUND CACHE FILEPATH "" FORCE)
    set(CMAKE_BUILD_WITH_INSTALL_RPATH TRUE)

    string(REPLACE "\\" "/" COMPILER_HOME $ENV{COMPILER_HOME})
endif()
```

With our Platform file (`VxWorks.cmake`) looking like this:
```Cmake
set(CMAKE_C_OUTPUT_EXTENSION .o)
set(CMAKE_CXX_OUTPUT_EXTENSION .o)
set(CMAKE_ASM_OUTPUT_EXTENSION .o)
set(CMAKE_STATIC_LIBRARY_PREFIX lib)
set(CMAKE_STATIC_LIBRARY_SUFFIX .r)
set(CMAKE_SHARED_LIBRARY_PREFIX )
set(CMAKE_SHARED_LIBRARY_SUFFIX .out)
set(CMAKE_EXECUTABLE_SUFFIX .exe)
```
I hope this works for you too. Please let me know, in case.

Many regards,
  Rainer

**Comment by rpoisel** [comment]
I have just seen, that toolchain files are supposed to be part of Paho. So, I will see to make a pull-request adding these required files for VxWorks too.

**Comment by cochicde** [comment]
Hi. thanks for the clarification. I used a similar VxWorks.cmake and toolchain files as yours. When compiling Paho, I was trying to use the same CMakeLists.txt files from Paho and just change there and add my flags when compiling. I had to avoid compiling the dynamic libraries to get a static one, since the dynamic was failing as told. I'm not very good at CMake so I'm not sure what's the best approach to add the VxWorks support there. 

Probably CMake is adding the flags when setting the target libraries that are giving me an error. I'll check if I can find where the -r comes from
